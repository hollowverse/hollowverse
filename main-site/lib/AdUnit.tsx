import { type ReactElement, useEffect, useMemo, useState } from 'react';

export interface InsertAdOptions {
  hbDisabled?: boolean;
}

type DestroyAd = () => void & {
  // The Ad ID that can be used with other API calls,
  // if the ad is inserted successfully.
  // The ID is local to the page view.
  id?: string;
};

// Assume that the pubg global instance has been initialized
// before the app is loaded.
interface Pubg {
  insertAd(
    deliveryId: string,
    options?: InsertAdOptions,
    div?: HTMLElement,
  ): DestroyAd;
  queue: { push(fn: () => unknown): void };
}

export interface AdProps {
  deliveryId: string;
  adOptions?: InsertAdOptions;
  className?: string;
}

function AdUnit({ deliveryId, adOptions, className }: AdProps): ReactElement {
  const [divEl, setDivEl] = useState<HTMLElement | null>(null);

  useEffect(() => {
    if (!divEl) {
      return;
    }

    const wnd = window as typeof window & { pubg: Pubg };

    // Inserting an ad is an async process,
    // and the component may be rendered multiple times before inserting the ad.
    // This flag is used to avoid the unnecessary inserting if the component has been unmounted.
    // It may avoid a lot of uninteresting errors that are sent to our servers.
    let ignore = false;
    let destroyAd: DestroyAd | undefined;

    wnd.pubg.queue.push(() => {
      if (!ignore) {
        destroyAd = wnd.pubg.insertAd(deliveryId, adOptions, divEl);
      }
    });

    return () => {
      ignore = true;

      wnd.pubg.queue.push(() => {
        if (destroyAd) {
          destroyAd();
        }
      });
    };
  }, [
    divEl,
    deliveryId,
    // adOptions should be considered as initial values and should not be changed on the fly.
  ]);

  // Importantly, the div should not be changed once created and should be memoized.
  // It may be attempted to use a ref instead of the callback,
  // but ref.current is updated outside React's flow,
  // and should not be used as a dependency of hooks.
  // Using a callback and a state can avoid such glitches.
  return useMemo(() => <div ref={setDivEl} className={className} />, []);
}

export default AdUnit;
