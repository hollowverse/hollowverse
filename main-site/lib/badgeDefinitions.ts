export const badgeDefinitions = {
  stardust: {
    name: 'Stardust',
    id: 104,
    stardustThreshold: 0,
  },

  theMoon: {
    name: 'The Moon',
    id: 114,
    stardustThreshold: 1,
  },

  halleysComet: {
    name: "Halley's Comet",
    id: 106,
    stardustThreshold: 5,
  },

  venus: {
    name: 'Venus',
    id: 107,
    stardustThreshold: 10,
  },

  theSun: {
    name: 'The Sun',
    id: 108,
    stardustThreshold: 25,
  },

  mercury: {
    name: 'Mercury',
    id: 109,
    stardustThreshold: 50,
  },

  hypotheticalMoonOfMercury: {
    name: 'Hypothetical Moon of Mercury',
    id: 116,
    stardustThreshold: 55,
  },

  mars: {
    name: 'Mars',
    id: 110,
    stardustThreshold: 100,
  },

  jupiter: {
    name: 'Jupiter',
    id: 111,
    stardustThreshold: 150,
  },

  saturn: {
    name: 'Saturn',
    id: 112,
    stardustThreshold: 200,
  },

  uranus: {
    name: 'Uranus',
    id: 113,
    stardustThreshold: 300,
  },

  neptune: {
    name: 'Neptune',
    id: 115,
    stardustThreshold: 500,
  },
};
