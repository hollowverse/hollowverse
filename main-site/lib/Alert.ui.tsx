/* This example requires Tailwind CSS v2.0+ */
import { ExclamationIcon, XCircleIcon, XIcon } from '@heroicons/react/solid';
import { MouseEventHandler, PropsWithChildren, ReactNode } from 'react';
import { c } from '~/lib/c';

export default function Alert(
  props: PropsWithChildren<{
    onDismiss?: MouseEventHandler<HTMLButtonElement>;
    color: 'yellow' | 'red';
  }>,
) {
  const { Icon, colors } = getColors();

  return (
    <div className={c('mx-2 rounded-md p-4', colors[0])}>
      <div className="flex">
        <div className="flex-shrink-0">
          <Icon className={c('h-5 w-5', colors[1])} aria-hidden="true" />
        </div>

        <div className="ml-3">{props.children}</div>

        {props.onDismiss && (
          <div className="ml-auto pl-3">
            <div className="-mx-1.5 -my-1.5">
              <button
                onClick={props.onDismiss}
                type="button"
                className={c(
                  'inline-flex rounded-md p-1.5 focus:outline-none focus:ring-2 focus:ring-offset-2',
                  colors[2],
                )}
              >
                <span className="sr-only">Dismiss</span>
                <XIcon className="h-5 w-5" aria-hidden="true" />
              </button>
            </div>
          </div>
        )}
      </div>
    </div>
  );

  function getColors() {
    if (props.color === 'red') {
      return {
        Icon: XCircleIcon,
        colors: [
          'bg-red-50',
          'text-red-400',
          'bg-red-50 text-red-500 hover:bg-red-100 focus:ring-red-600 focus:ring-offset-red-50',
        ],
      } as const;
    }

    return {
      Icon: ExclamationIcon,
      colors: [
        'bg-yellow-50',
        'text-yellow-400',
        'bg-yellow-50 text-yellow-500 hover:bg-yellow-100 focus:ring-yellow-600 focus:ring-offset-yellow-50',
      ],
    } as const;
  }
}
