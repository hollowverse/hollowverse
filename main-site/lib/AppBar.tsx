import { ReactNode, useState } from 'react';
import { FaBars, FaSearch, FaTimes } from 'react-icons/fa';
import { c } from '~/lib/c';
import { Card } from '~/lib/Card.ui';
import { Link } from '~/lib/Link';
import { LovelyTopBorder } from '~/lib/LovelyTopBorder';
import { useLocationHref } from '~/lib/useLocationHref';
import { useUser } from '~/lib/useUser';

export function Logo(props: { className?: string } = {}) {
  return (
    <div
      className={c('flex w-full items-center default:text-xl', props.className)}
    >
      <Link passHref href="/">
        <a
          id="logo"
          className="lovely-gradient w-fit cursor-pointer select-none bg-clip-text font-extrabold uppercase tracking-tight text-transparent"
        >
          Hollowverse
        </a>
      </Link>
    </div>
  );
}

export function AppBar() {
  const [open, setOpen] = useState(false);
  const href = useLocationHref();
  const user = useUser();

  return (
    <Card topBorder={false} className="font-normal">
      <LovelyTopBorder />

      <nav
        role="navigation"
        aria-label="Main Navigation"
        className="h-container flex flex-col gap-3 py-3 px-5"
      >
        <div className="flex items-center justify-between default:flex-row default:gap-2">
          <Logo />

          <div className="flex gap-3">
            <Link
              href={{
                pathname: '/~search',
                query: { local: true },
              }}
              passHref
            >
              <a
                id="search-link"
                className="flex items-center gap-1 p-1.5 font-normal text-neutral-400"
              >
                <FaSearch className="h-6 w-6" />
                <span className="hidden text-base uppercase md:block">
                  Search
                </span>
              </a>
            </Link>

            <button
              className="flex items-center gap-1 p-1.5 font-normal text-neutral-400"
              onClick={() => setOpen(!open)}
            >
              {open ? (
                <FaTimes className="h-6 w-6" />
              ) : (
                <FaBars className="h-6 w-6" />
              )}
              <span className="hidden text-base font-normal uppercase text-neutral-400 md:block">
                Menu
              </span>
            </button>
          </div>
        </div>

        {open && (
          <div className="flex flex-col items-end gap-3 border-t py-5">
            <a
              href="https://forum.hollowverse.com"
              className="h-link underline"
            >
              Forum
            </a>

            {(user && (
              <a href="/~logout" className="h-link underline">
                Sign-out
              </a>
            )) || (
              <a
                className="h-link underline"
                href={`/api/login?redirect=${encodeURI(href)}`}
              >
                Sign-in
              </a>
            )}
          </div>
        )}
      </nav>
    </Card>
  );
}
