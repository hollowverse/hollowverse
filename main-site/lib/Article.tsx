/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState } from 'react';
import { FaRegEdit } from 'react-icons/fa';
import { c } from '~/lib/c';
import { CelebPageMainProps } from '~/lib/celebPageMain.getStaticProps';
import { TitledCard } from '~/lib/TitledCard.ui';
import s from '~/styles/styles.module.scss';

export const Article = (props: CelebPageMainProps) => {
  const [showFootnotes, setShowFootnotes] = useState(false);

  return (
    <article
      className="flex flex-col gap-5"
      onClick={(e) => {
        if (
          (e.target as Element)?.parentElement?.classList.contains(
            'footnote-ref',
          )
        ) {
          setShowFootnotes(true);
        }
      }}
    >
      <TitledCard
        titledContentProps={{
          title: (
            <span className="flex gap-3">
              Wiki{' '}
              <a
                href={props.celeb.wiki!}
                className="h-link flex items-center text-base font-normal"
              >
                <span className="flex items-center gap-1 ">
                  <FaRegEdit /> Edit
                </span>
              </a>
            </span>
          ),
        }}
      >
        <div className={s.Article}>
          <div
            id="editorial"
            className={c('break-normal p-5 pb-0 leading-relaxed', {
              'show-footnotes': showFootnotes,
            })}
            dangerouslySetInnerHTML={{ __html: props.wiki.content }}
          />

          <button
            className="px-5 pb-5 text-sm text-gray-500"
            onClick={() => setShowFootnotes(!showFootnotes)}
          >
            ({showFootnotes ? 'Hide footnotes' : 'Show footnotes'})
          </button>
        </div>
      </TitledCard>
    </article>
  );
};
