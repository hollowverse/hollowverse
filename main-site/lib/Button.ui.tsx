import { c } from '~/lib/c';
import { ReactElementProps } from '~/shared/lib/types';

export function Button(props: ReactElementProps<'button'>) {
  return (
    <button
      {...props}
      className={c(
        'ml-3 inline-flex justify-center rounded-md border border-transparent bg-purple-600 py-2 px-10 text-sm font-medium text-white shadow-sm hover:bg-purple-700 focus:outline-none focus:ring-2 focus:ring-purple-500 focus:ring-offset-2',
        props.className,
      )}
    />
  );
}
