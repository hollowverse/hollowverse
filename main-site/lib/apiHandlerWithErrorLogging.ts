import { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
import { log } from '~/shared/lib/log';

export function apiHandlerWithErrorLogging(
  apiRouteName: string,
  fn: NextApiHandler<any>,
) {
  return async function (req: NextApiRequest, res: NextApiResponse) {
    try {
      return await fn(req, res);
    } catch (e: any) {
      log('error', e, [apiRouteName]);
      throw e;
    }
  };
}
