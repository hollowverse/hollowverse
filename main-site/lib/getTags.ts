import { getYear } from 'date-fns';
import {
  differenceWith,
  filter,
  first,
  flatten,
  groupBy,
  isEmpty,
  last,
  mapValues,
  pullAt,
  reverse,
  sortBy,
  toPairs,
  uniqBy,
} from 'lodash-es';
import { parseDate } from '~/lib/date';
import { Fact, Tag } from '~/lib/groq/fact.projection';
import { OrderOfIssues } from '~/lib/groq/orderOfIssues.projection';

export function sortTags(tags: Tag[], orderOfIssues: OrderOfIssues) {
  const noDupes = uniqBy(
    tags,
    (t) => `${t.tag.name}-is-low-confidence-${t.isLowConfidence}`,
  );

  // Remove lowConfidence tags if a regular one exists
  const noUnnecessaryLowConfidence = noDupes.filter((t) => {
    if (!t.isLowConfidence) {
      return true;
    }

    const regularExists = noDupes.some((t2) => {
      return t2.tag.name === t.tag.name && !t2.isLowConfidence;
    });

    if (regularExists) {
      return false;
    }

    return true;
  });

  const diff = differenceWith(
    noUnnecessaryLowConfidence,
    orderOfIssues,
    (t, i) => t.tag.issue.name === i,
  );

  const intersection = filter(noUnnecessaryLowConfidence, (t) => {
    return orderOfIssues.includes(t.tag.issue.name);
  });

  intersection.sort(function (a, b) {
    return (
      orderOfIssues.indexOf(a.tag.issue.name) -
      orderOfIssues.indexOf(b.tag.issue.name)
    );
  });

  return [...intersection, ...diff];
}

type TagPair = [string, Tag[]];

/**
group the tags by the year. then sort by the most recent years first. but the
problem is, it doesn't look very good when we have a bunch of years, with each
one having just one tag.

so here's how i solved that problem:

starting from the oldest year (i.e. from the bottom), if there are less than 5
facts in that year, roll those up with the next year (so instead of a single
year now, we have a time period yearX-yearY), then if that time period still has
less than 5 facts, those too roll them up with the next year. keep rolling up
until the period has 5 or more tags.

now there's another issue..for many celebs, because they don't have many tags
per year, when you roll up their tags, they end up with a single time period,
2022-2019 four tags. showing a timeline with a single time period doesn't look
good. so in that case, i simply hide the timeline and just display the tags
without the date.
 */
export function rollUp(tagPairs: TagPair[]) {
  const pending: TagPair[] = [];

  for (let i = tagPairs.length - 1; i >= 0; i--) {
    const tagPair = tagPairs[i];
    const [title, tags] = tagPair;

    if (!isEmpty(pending)) {
      const preNewTitle = `${title}-${pending[0][0]}`.split('-');
      const newTitle = `${first(preNewTitle)}-${last(preNewTitle)}`;
      tagPair[0] = newTitle;
      tagPair[1] = [...tags, ...pending[0][1]];
      pending.length = 0;
    }

    if (tagPair[1].length < 5 && i !== 0) {
      pending.push(tagPair);
      pullAt(tagPairs, i);
    }
  }

  return tagPairs;
}

export function getTags(facts: Fact[], orderOfIssues: OrderOfIssues) {
  const factsByYear = groupBy(facts, (f) => {
    return getYear(parseDate(f.date));
  });
  const tagsByYearObj = mapValues(factsByYear, (facts) => {
    return flatten(facts.map((f) => f.tags));
  });

  const tagsByYearUnsorted = toPairs(tagsByYearObj);
  const tagsByYear = reverse(sortBy(tagsByYearUnsorted, (pair) => pair[0]));

  rollUp(tagsByYear);

  const withSortedTagClouds = tagsByYear.map((tby) => {
    return [tby[0], sortTags(tby[1], orderOfIssues)] as TagPair;
  });

  return withSortedTagClouds;
}
