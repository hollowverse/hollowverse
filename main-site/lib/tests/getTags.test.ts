import { rollUp, sortTags } from '~/lib/getTags';
import { mockOrderOfIssues } from '~/lib/tests/mockOrderOfIssues';

test('sortTags to remove dupes', () => {
  expect(
    sortTags(
      [
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
      ],
      mockOrderOfIssues,
    ),
  ).toMatchSnapshot();
});

test('sortTags to remove unnecessary lowConfidence tags', () => {
  expect(
    sortTags(
      [
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: true,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
      ],
      mockOrderOfIssues,
    ),
  ).toMatchSnapshot();

  expect(
    sortTags(
      [
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: true,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },

        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: true,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: true,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'foobar',
          },
        },
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Environmentalist',
          },
        },
      ],
      mockOrderOfIssues,
    ),
  ).toMatchSnapshot();
});

test('sortTags should sort by order of issues', () => {
  expect(
    sortTags(
      [
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'LGBT',
            },
            name: 'Supports LGBT',
          },
        },
        {
          _id: 'id',
          isBackground: null,
          isLowConfidence: null,
          tag: {
            issue: {
              name: 'Environment',
            },
            name: 'Pro Nuclear Power',
          },
        },
        {
          _id: 'id',
          isBackground: false,
          isLowConfidence: true,
          tag: {
            issue: {
              name: 'Religion',
            },
            name: 'Not Religious',
          },
        },
      ],
      mockOrderOfIssues,
    ),
  ).toMatchSnapshot();
});

describe('rollUp', () => {
  describe('simple case', () => {
    expect(
      rollUp([
        ['2020', ['a', 'b', 'c']],
        ['2019', ['d', 'e', 'f']],
      ] as any),
    ).toEqual([['2020-2019', ['a', 'b', 'c', 'd', 'e', 'f']]]);
  });

  describe('never less than five per period', () => {
    expect(
      rollUp([
        ['2020', ['a', 'b', 'c', 'd', 'e', 'f']],
        ['2019', ['d', 'e', 'f']],
      ] as any),
    ).toEqual([['2020-2019', ['a', 'b', 'c', 'd', 'e', 'f', 'd', 'e', 'f']]]);
  });

  describe('last and previous years are short', () => {
    expect(
      rollUp([
        ['2020', ['a', 'b', 'c']],
        ['2019', ['d', 'e', 'f']],
        ['2018', ['g', 'h', 'i', 'j', 'k']],
      ] as any),
    ).toEqual([
      ['2020-2019', ['a', 'b', 'c', 'd', 'e', 'f']],
      ['2018', ['g', 'h', 'i', 'j', 'k']],
    ]);
  });

  describe('only last year is short', () => {
    expect(
      rollUp([
        ['2020', ['a', 'b']],
        ['2019', ['e', 'f', 'g', 'h', 'i']],
        ['2018', ['g', 'h', 'i', 'j', 'k']],
      ] as any),
    ).toEqual([
      ['2020', ['a', 'b']],
      ['2019', ['e', 'f', 'g', 'h', 'i']],
      ['2018', ['g', 'h', 'i', 'j', 'k']],
    ]);
  });

  describe('a period can grow larger than might be expected', () => {
    expect(
      rollUp([
        ['2022', ['a', 'b', 'c', 'd', 'e', 'f']],
        ['2021', ['g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p']],
        ['2020', ['q']],
        ['2018', ['r', 's']],
        ['2017', ['t']],
      ] as any),
    ).toEqual([
      ['2022', ['a', 'b', 'c', 'd', 'e', 'f']],
      [
        '2021-2017',
        ['g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't'],
      ],
    ]);
  });
});
