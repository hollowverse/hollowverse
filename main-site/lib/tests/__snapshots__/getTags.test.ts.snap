// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortTags should sort by order of issues 1`] = `
Array [
  Object {
    "isBackground": false,
    "isLowConfidence": true,
    "tag": Object {
      "issue": Object {
        "name": "Religion",
      },
      "name": "Not Religious",
    },
  },
  Object {
    "isBackground": null,
    "isLowConfidence": null,
    "tag": Object {
      "issue": Object {
        "name": "LGBT",
      },
      "name": "Supports LGBT",
    },
  },
  Object {
    "isBackground": null,
    "isLowConfidence": null,
    "tag": Object {
      "issue": Object {
        "name": "Environment",
      },
      "name": "Pro Nuclear Power",
    },
  },
]
`;

exports[`sortTags to remove dupes 1`] = `
Array [
  Object {
    "isBackground": null,
    "isLowConfidence": null,
    "tag": Object {
      "issue": Object {
        "name": "Environment",
      },
      "name": "Environmentalist",
    },
  },
]
`;

exports[`sortTags to remove unnecessary lowConfidence tags 1`] = `
Array [
  Object {
    "isBackground": null,
    "isLowConfidence": null,
    "tag": Object {
      "issue": Object {
        "name": "Environment",
      },
      "name": "Environmentalist",
    },
  },
]
`;

exports[`sortTags to remove unnecessary lowConfidence tags 2`] = `
Array [
  Object {
    "isBackground": null,
    "isLowConfidence": null,
    "tag": Object {
      "issue": Object {
        "name": "Environment",
      },
      "name": "Environmentalist",
    },
  },
  Object {
    "isBackground": null,
    "isLowConfidence": true,
    "tag": Object {
      "issue": Object {
        "name": "Environment",
      },
      "name": "foobar",
    },
  },
]
`;
